autofit/graphical/expectation_propagation/ep_mean_field.py

```diff
@@ -223,3 +326,165 @@ def __repr__(self) -> str:
+    this class encode the EP mean-field approximation to a factor graph
+
+
+    Attributes
+    ----------
+    factor_graph: FactorGraph
+        the base factor graph being approximated
+
+    factor_mean_field: Dict[Factor, MeanField]
+        the mean-field approximation for each factor in the factor graph
+
+    mean_field: MeanField
+        the mean-field approximation of the full factor graph
+        i.e. the product of the factor mean-field approximations
+
+    variables: Set[Variable]
+        the variables of the approximation
+
+    deterministic_variables: Set[Variable]
+        the deterministic variables
+
+    log_evidence: float
+        the approximate log evidence of the approximation
+
+    is_valid: bool
+        returns whether the factor mean-field approximations are all valid
+
+    Methods
+    -------
+    from_approx_dists(factor_graph, approx_dists)
+        create a EPMeanField object from the passed factor_graph
+        using approx_dists to initialise the factor mean-field approximations
+
+    factor_approximation(factor)
+        create the FactorApproximation for the factor
+
+    project_factor_approx(factor_approximation)
+        given the passed FactorApproximation, return a new `EPMeanField`
+        object encoding the updated mean-field approximation
+    """
+
+    def __init__(
+        self,
+        factor_graph: FactorGraph,
+        factor_mean_field: Dict[Factor, MeanField],
+        factor_rescale: Dict[Factor, Dict[Variable, float]],
+        factor_subset_factor: Dict[Factor, Factor],
+        ep_mean_field: EPMeanField,
+        plates_index: Dict[Plate, List[int]],
+    ):
+        super().__init__(factor_graph, factor_mean_field)
+
+        self._factor_rescale = factor_rescale
+        self._factor_subset_factor = factor_subset_factor
+        self._ep_mean_field = ep_mean_field
+        self._plates_index = plates_index
+
+    
+    @property 
+    def parameters(self):
+        return {
+            "factor_graph": self._factor_graph, 
+            "factor_mean_field": self._factor_mean_field,
+            "factor_rescale": self._factor_rescale,
+            "factor_subset_factor": self._factor_subset_factor,
+            "ep_mean_field": self._ep_mean_field,
+            "plates_index": self._plates_index,
+        }
+
+    @property 
+    def plates_index(self):
+        return self._plates_index.copy()
+
+    @property 
+    def factor_rescale(self):
+        return self._factor_rescale.copy()
+
+    @property 
+    def factor_subset_factor(self):
+        return self._factor_subset_factor.copy()
+
+    def project_mean_field(
+        self,
+        new_dist: MeanField,
+        factor_approx: FactorApproximation,
+        delta: float = 1.0,
+        status: Status = Status(),
+    ) -> Tuple["EPMeanField", Status]:
+
+        factor_mean_field = self.factor_mean_field
+        
+        # We're fitting the full factor_dist, not the subset factor_dist
+        rescale1 = {
```

Comment:
Why is this rescaling necessary?

Reply:
*waves hands at maths*

The rescaling is necessary as we're effectively factorising the subfactor out of the factor, so we have to account for that when calculating the cavity and factor distributions. 

I'm not 100% confident that the rescaling has to be done per variable as opposed to per factor but the tests kinda work.